R/tar_renv.R

```diff
@@ -25,35 +26,55 @@
 #'   library. See <https://rstudio.github.io/renv/index.html> for
 #'   more information.
 #' @return Nothing, invisibly.
-#' @inheritParams tar_option_set
+#' @inheritParams tar_prune
+#' @param extras Character vector of additional packages to declare as
+#'   project dependencies.
 #' @param path Character of length 1, path to the script file to
 #'   populate with `library()` calls.
-#' @param ask Logical, whether to ask before writing if `_packages.R`
-#'   already exists. If `NULL`, defaults to `Sys.getenv("TAR_ASK")`.
-#'   (Set to `"true"` or `"false"` with `Sys.setenv()`).
-#'   If `ask` and the `TAR_ASK` environment variable are both
-#'   indeterminate, defaults to `interactive()`.
 #' @examples
 #' tar_dir({
-#'   tar_option_set(packages = c("tibble", "qs"))
+#'   tar_script({
+#'     tar_option_set(packages = c("tibble", "qs"))
+#'     tar_pipeline()
+#'   })
 #'   tar_renv()
 #'   readLines("_packages.R")
 #' })
 tar_renv <- function(
-  packages = targets::tar_option_get("packages"),
+  extras = c("clustermq", "future", "rstudioapi", "visNetwork"),
   path = "_packages.R",
-  ask = NULL
+  callr_function = callr::r,
+  callr_arguments = list()
 ) {
-  assert_chr(path)
-  assert_scalar(path)
-  if (!tar_should_overwrite(ask, path)) {
-    # Covered in tests/interactive/test-tar_renv.R. # nolint
-    return(invisible()) #nocov
-  }
+  assert_target_script()
+  assert_chr(extras, "extras must be a character vector")
+  assert_chr(path, "path must have type character")
+  assert_scalar(path, "path must have length 1")
+  assert_callr_function(callr_function)
+  assert_list(callr_arguments, "callr_arguments must be a list.")
+  callr_outer(
+    targets_function = tar_renv_inner,
+    targets_arguments = list(extras = extras, path = path),
+    callr_function = callr_function,
+    callr_arguments = callr_arguments
+  )
+  invisible()
+}
+
+tar_renv_inner <- function(pipeline, extras, path) {
+  pipeline_validate_lite(pipeline)
+  option_pkgs <- tar_option_get("packages")
+  targets_pkgs <- pipeline_get_packages(pipeline)
+  option_format <- store_get_packages(store_init(tar_option_get("format")))
+  targets_formats <- pipeline_get_formats(pipeline)
+  targets_formats <- map(targets_formats, ~store_get_packages(store_init(.)))
+  targets_formats <- unlist(targets_formats)
+  all_pkgs <- unique(c(option_pkgs, targets_pkgs, option_format,
+                       targets_formats, extras))
```

Comment:
I prefer [tidyverse-style indenting](https://style.tidyverse.org/pipes.html#long-lines-2) for calls over 80 characters:

```r
all_pkgs <- unique(
  c(
    option_pkgs,
    targets_pkgs,
    option_format,
    extras
  )
)
```